### PR TITLE
pow_dense

### PR DESCRIPTION
**Pull request by MrRobot2211**
This implements pow and creates a class that should be inside QuTiP proper to do the testing.

We can afterwards add the test class to qutip and inherit here in future PRs.

This in pursuit of having the most functionality available on the "first" release.

This is ready to merge after extra_functions

### CODE REVIEW

File: src/qutip_cupy/dense_functions.py [hunk 1]
```diff
@@ -29,3 +29,12 @@ def trace_cupydense(cp_arr):
     # @TODO: whnen qutip allows it we should remove this call to item()
     # as it takes a time penalty commmunicating data from GPU to CPU.
     return cp.trace(cp_arr._cp).item()
+
+
+def pow_cupydense(cp_arr, n):
+    if cp_arr.shape[0] != cp_arr.shape[1]:
+        raise TypeError("matrix power only works with square matrices")
```
Comment (hodgestar) [inline src/qutip_cupy/dense_functions.py:36]:
```suggestion
        raise ValueError("matrix power only works with square matrices")
```
The type is correct, but the shape is not.

File: tests/test_mathematics.py [hunk 2]
```diff
@@ -86,3 +88,37 @@ class TestTranspose(test_tools.TestTranspose):
...
+
+    def op_numpy(self, matrix, scalar):
+
+        return np.linalg.matrix_power(matrix, scalar)
+
+    shapes = [
+        (pytest.param((1, 1),),),
+        (pytest.param((5, 5),),),
+        (pytest.param((10, 10),),),
+    ]
```
Comment (hodgestar) [inline tests/test_mathematics.py:106]: We should add tests for the exception case too -- i.e. `bad_shapes = [...]` and the suitable error test method.

**Review by hodgestar — CHANGES_REQUESTED** [review]

**Comment by AGaliciaMartinez** [comment]
The class `PowTest` is now merged in dev.major. :)

**Comment by MrRobot2211** [comment]
@hodgestar I think this is ready then, I'll inherit from the newly minted QuTiP tests later on.

**Review by hodgestar — APPROVED** [review]

File: tests/test_mathematics.py [hunk 3]
```diff
@@ -88,8 +90,59 @@ class TestTranspose(test_tools.TestTranspose):
...
+        with pytest.raises(ValueError):
+            op(data_m(), scalar)
+
+
 class TestProject(test_tools.TestProject):
 
     specialisations = [
         pytest.param(cdf.project_cupydense, CuPyDense, CuPyDense),
-    ]
+    ]
```
Comment (hodgestar) [inline tests/test_mathematics.py:148]:
```suggestion
    ]

```

File: src/qutip_cupy/dense_functions.py [hunk 4]
```diff
@@ -43,4 +52,4 @@ def project_cupydense(state):
     elif state.shape[0] == 1:
         return CuPyDense._raw_cupy_constructor(cp.outer(state.adjoint()._cp, state._cp))
     else:
-        raise ValueError("state must be a ket or a bra.")
+        raise ValueError("state must be a ket or a bra.")
```
Comment (hodgestar) [inline src/qutip_cupy/dense_functions.py:55]:
```suggestion
        raise ValueError("state must be a ket or a bra.")
```

**Review by hodgestar — APPROVED** [review]
I had to resolve some merge conflicts. Approved again. :)

**Comment by hodgestar** [comment]
@MrRobot2211 Approved! Happy for you to merge once the test run passes.

File: src/qutip_cupy/dense_functions.py [hunk 5]
```diff
@@ -31,6 +31,15 @@ def trace_cupydense(cp_arr):
...
 
+def pow_cupydense(cp_arr, n):
+    if cp_arr.shape[0] != cp_arr.shape[1]:
+        raise ValueError("matrix power only works with square matrices")
+
+    out_arr = cp.linalg.matrix_power(cp_arr._cp, n)
+
+    return CuPyDense._raw_cupy_constructor(out_arr)
+
+
```
Comment (hodgestar) [inline src/qutip_cupy/dense_functions.py:42]:
```suggestion

```

**Comment by hodgestar** [comment]
@MrRobot2211 Could you fix up the silly pep 8 errors resulting from my merging with GitHub's merge interface? Apologies for leaving you with the mess but I am stepping away from my computer now and don't have the branch checked out to do the clean up quickly.

**Pull request merged**